diffrax/_solver/srk.py

```diff
@@ -0,0 +1,633 @@
+    then `bW` and `bH` should be arrays of shape `(k,)`, `cW` and `cH` should be `None`,
+    and `aW` and `aH` should be lists of arrays of the same shape as `a`.
+"""
+
+
+class AbstractSRK(AbstractStratonovichSolver):
+    r"""A general Stochastic Runge-Kutta method.
+
+    The second term in the MultiTerm must be a `ControlTerm` with
+    `control=VirtualBrownianTree`. Depending on the Butcher tableau, the
+    `VirtualBrownianTree` may need to be initialised with 'levy_area="space-time"'
+    or 'levy_area="space-time-time"'.
+
+    Given the Stratonovich SDE
+    $dX_t = f(t, X_t) dt + g(t, X_t) \circ dW_t$
+
+    We construct the SRK with $s$ stages as follows:
+
+    $y_{n+1} = y_n + h \Big(\sum_{j=1}^s b_j f_j \Big)
+    + W_n \Big(\sum_{i=1}^{j-1} b^W_j g_i \Big)
+    + H_n \Big(\sum_{i=1}^{j-1} b^H_j g_i \Big)$
+
+    $f_j = f(t_0 + c_j h , z_j)$
+
+    $g_j = g(t_0 + c_j h , z_j)$
+
+    $z_j = y_n + h \Big(\sum_{i=1}^{j-1} a_{j,i} f_i \Big)
+    + W_n \Big(\sum_{i=1}^{j-1} a^W_{j,i} g_i \Big)
+    + H_n \Big(\sum_{i=1}^{j-1} a^H_{j,i} g_i \Big)$
+
+    where $W_n = W_{t_n, t_{n+1}}$ is the increment of the Brownian motion and
+    $H_n = H_{t_n, t_{n+1}}$ is its corresponding space-time Lévy Area.
+    A similar term can also be added for the space-time-time Lévy area, K.
+
+    In the special case, when the SDE has additive noise, i.e. when g is
+    indepnedent of y (but can still depend on t), then the SDE can be written as
+    $dX_t = f(t, X_t) dt + g(t) \, dW_t$, and we can simplify the above to
+
+    $y_{n+1} = y_n + h \Big(\sum_{j=1}^s b_j k_j \Big) + g(t_n) \, (b^W
+    \, W_n + b^H \, H_n)$
+
+    $f_j = f(t_n + c_j h , z_j)$
+
+    $z_j = y_n + h \Big(\sum_{i=1}^{j-1} a_{j,i} f_i \Big) + g(t_n)
+    \, (c^W_j W_n + c^H_j H_n)$
+
+    When g depends on t, we need to add a correction term to $y_{n+1}$ of
+    the form $(g(t_{n+1}) - g(t_n)) \, (1/2 W_n - H_n)$.
+
+    The coefficients are provided in the `StochasticButcherTableau`.
+    """
+
+    term_structure = MultiTerm[tuple[ODETerm, AbstractTerm]]
+    interpolation_cls = LocalLinearInterpolation
+    tableau: StochasticButcherTableau
+
+    # Indicates the type of Levy area used by the solver.
+    # The BM must generate at least this type of Levy area, but can generate
+    # more. E.g. if the solver uses space-time Levy area, then the BM generates
+    # space-time-time Levy area as well that is fine. The other way around would
+    # not work. This is mostly an easily readable indicator so that methods know
+    # what kind of BM to use.
+    minimal_levy_area: MinimalLevyArea
+
+    def __init__(self):
+        if self.tableau.bK is not None:
+            self.minimal_levy_area: MinimalLevyArea = "space-time-time"
+        elif self.tableau.bH is not None:
+            self.minimal_levy_area: MinimalLevyArea = "space-time"
+        else:
+            self.minimal_levy_area: MinimalLevyArea = ""
+
+    def init(
+        self,
+        terms: term_structure,
```

Comment:
Oh -- I've never seen someone do a type annotation like this before. Does this really pass pyright?

Reply:
It does pass yes. I thought that's why term_structure was declared in the first place. Makes it easier to modify. But if you prefer I can revert it to the original.

Reply:
No strong feelings, to be honest.
(`term_structure` is declared for use at runtime -- `diffeqsolve` checks that the provided `terms` match the structure required by the solver.)